### PR TITLE
Replace simplejson with rapidjson

### PR DESCRIPTION
**Pull request by citrus-it**
I've only done limited testing on OpenIndiana for this but I wanted to share a change that we recently made in OmniOS IPS, which is to replace the `simplejson` library with `rapidjson`.  We also centralised the JSON parsing and schema checking into a new `pkg.json` module which makes it easier to drop in different JSON implementations if necessary. While rapidjson does provide schema validation, it didn't work in our testing so jsonschema is still used for that.

This followed some discussions at FOSDEM (thanks @Toasterson) about memory usage in IPS and in particular related to JSON parsing.

We evaluated a number of libraries including the standard python3 json module, the existing simplejson, rapidjson, orjson, ultrajson. We found that rapidjson uses about 20% less memory overall and is around 15% faster than simplejson for combined tasks such as a `pkg refresh --full` or `pkg list -a`. The other libraries were the same or worse than simplejson (and some are really hard to build on illumos too).

Here's an extract of the memory benchmarking showing the total RSS used when loading and parsing the OpenIndiana summary catalogue file, which is 77MiB as of today:

```
77.0M catalog.summary.C

rapidjson       233124K
orjson          363556K
stdlib          363672K
simplejson      364156K
hyperjson       413920K
```

and here are the testsuite results (obviously this requires `rapidjson-35` for which there is a draft PR at https://github.com/OpenIndiana/oi-userland/pull/5570 )

```
# Ran 1132 tests in 1807.478s - skipped 2 tests.

FAILED (successes=1129, failures=3, errors=0, mismatches=2)

======================================================================
BASELINE MISMATCH: The following results didn't match the baseline.
----------------------------------------------------------------------
cli.t_pkg_sysrepo.py TestSysrepo.test_disabled_origins: fail
cli.t_pkg_depotd.py TestDepotOutput.test_0_depot_bui_output: fail
======================================================================
```

### CODE REVIEW

**Comment by Toasterson** [comment]
Thank you @citrus-it for adding this. Considering that we are also unserializing the JSON into Python Classes the total usage will be much higher. Did you by any chance do a count which packages are filling up the Catalog JSON files? Last time I looked it was mostly userland-incorporation

**Comment by citrus-it** [comment]
The benchmarks include the de-serialisation.

With the benchmark version of the code uncommented, you can see which parts are causing most memory usage: (this is the rapidjson code)

```
% python3 =pkg -Djson=1 list -g http://pkg.openindiana.org/hipster
JSON Mem +0.07 MiB - Time 0.00s - load(0.00 MiB /var/pkg/state/installed/catalog.attrs) =
JSON Mem +1.17 MiB - Time 0.01s - load(0.16 MiB /var/pkg/state/installed/catalog.base.C) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB 3) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB 3) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /var/pkg/state/avoid_set) =
JSON Mem +0.02 MiB - Time 0.00s - loads(1166) =

Refreshing catalog 1/2 hipster-encumbered
JSON Mem +0.02 MiB - Time 0.00s - load(0.00 MiB /data/omnios-build/omniosorg/bloody/extra/tmp.repo/publisher/extra.omnios/catalog/catalog.attrs) =
JSON Mem +0.02 MiB - Time 0.00s - load(0.00 MiB /data/omnios-build/omniosorg/bloody/_repo/publisher/omnios/catalog/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.attrs) =
JSON Mem +0.04 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fedb84f90>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/tmp3h33v9vh/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fedb99850>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fea05ee50>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/tmpz4bbowh7/catalog.attrs) =
JSON Mem +0.07 MiB - Time 0.00s - load(0.02 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.base.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fea0a3490>) = None
JSON Mem +0.29 MiB - Time 0.00s - load(0.17 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.dependency.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe98dda50>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.12 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.summary.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe98e9450>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fedb99850>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.02 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.base.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fea094d10>) = None
JSON Mem +0.11 MiB - Time 0.00s - load(0.17 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.dependency.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe9905650>) = None
JSON Mem +0.30 MiB - Time 0.00s - load(0.12 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.summary.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe9dcc950>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.02 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.base.C) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.02 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.base.C) =
JSON Mem +0.04 MiB - Time 0.00s - load(0.17 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.dependency.C) =
JSON Mem +0.27 MiB - Time 0.00s - load(0.12 MiB /tmp/tmpqmkjbtqh/18446740225052517392/origins/0f1ac31412052e780a0f3388821fdb0b897014d6/catalog.summary.C) =
JSON Mem +0.00 MiB - Time 0.00s - dump(0.03 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.base.C) = None
JSON Mem +0.00 MiB - Time 0.00s - dump(0.16 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.dependency.C) = None
JSON Mem +0.00 MiB - Time 0.00s - dump(0.12 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.summary.C) = None
JSON Mem +0.00 MiB - Time 0.00s - dump(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.attrs) = None

Refreshing catalog 2/2 hipster-encumbered
JSON Mem +0.00 MiB - Time 0.00s - load(0.00 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.attrs) =

Refreshing catalog 2/2 openindiana.org
JSON Mem +0.00 MiB - Time 0.00s - load(0.06 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.attrs) =
JSON Mem +0.02 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fedba8090>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.06 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/tmpkenyn04q/catalog.attrs) =
JSON Mem +75.88 MiB - Time 0.24s - load(15.05 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.base.C) =
JSON Mem +0.27 MiB - Time 0.18s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe34ff110>) = None
JSON Mem +287.93 MiB - Time 1.37s - load(133.35 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.dependency.C) =
JSON Mem +0.27 MiB - Time 1.17s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fd1815b50>) = None
JSON Mem +204.18 MiB - Time 0.96s - load(77.03 MiB /tmp/download-3293-xge28h5g/incoming-3293/catalog.summary.C) =
JSON Mem +0.00 MiB - Time 0.72s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fd7eb2310>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.06 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fea140050>) = None
JSON Mem +73.18 MiB - Time 0.26s - load(15.05 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.base.C) =
JSON Mem +0.00 MiB - Time 0.18s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fe3a01850>) = None
JSON Mem +283.57 MiB - Time 1.46s - load(133.35 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.dependency.C) =
JSON Mem +0.27 MiB - Time 1.17s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fcb4096d0>) = None
JSON Mem +208.50 MiB - Time 1.19s - load(77.03 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.summary.C) =
JSON Mem +0.28 MiB - Time 0.71s - dump(<pkg.catalog._JSONWriter object at 0xfffffc7fb9af5cd0>) = None
JSON Mem +0.00 MiB - Time 0.00s - load(0.06 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.attrs) =
JSON Mem +72.23 MiB - Time 0.25s - load(15.05 MiB /tmp/tmpqmkjbtqh/18446740225052517200/origins/1e037461fe272285444a1caec32ee8a269d3a9d7/catalog.base.C) =

Refreshing catalog 2/2 openindiana.org
JSON Mem +0.00 MiB - Time 0.00s - load(0.06 MiB /tmp/tmpqmkjbtqh/18446740225052517200/catalog/catalog.attrs) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.03 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.base.C) =
JSON Mem +72.44 MiB - Time 0.24s - load(15.05 MiB /tmp/tmpqmkjbtqh/18446740225052517200/catalog/catalog.base.C) =
JSON Mem +0.38 MiB - Time 0.00s - load(0.17 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.dependency.C) =
JSON Mem +0.31 MiB - Time 0.00s - load(0.12 MiB /tmp/tmpqmkjbtqh/18446740225052517392/catalog/catalog.summary.C) =
JSON Mem +282.90 MiB - Time 1.39s - load(133.35 MiB /tmp/tmpqmkjbtqh/18446740225052517200/catalog/catalog.dependency.C) =
JSON Mem +208.75 MiB - Time 1.65s - load(77.03 MiB /tmp/tmpqmkjbtqh/18446740225052517200/catalog/catalog.summary.C) =
JSON Mem +0.00 MiB - Time 0.00s - load(0.39 MiB /var/pkg/state/installed/catalog.dependency.C) =
```

**Comment by citrus-it** [comment]
Most of the memory usage on OpenIndiana comes from the dependency catalogue, and in particular from the consolidations:

The output is `package name`, `no. of versions`, `no. of dependency actions across all versions`

```
omnios% jq '."openindiana.org" | to_entries | .[] | [.key, (.value | length), ([.value | .[] | .actions | length] | add)] | map(tostring) | join(" ")' < ./catalog.dependency.C | sort -n +2 | tail -10
"server_install 7 2733"
"system/file-system/zfs 175 2767"
"system/test/libdrm 373 2984"
"system/library 175 3117"
"mate_install 7 4105"
"SUNWcs 175 5079"
"developer/opensolaris/osnet 175 5379"
"consolidation/osnet/osnet-redistributable 175 81392"
"consolidation/osnet/osnet-incorporation 175 85417"
"consolidation/userland/userland-incorporation 339 679220"
```

**Comment by danmcd** [comment]
Probably not relevant here, but I was informed of a small C implementation called Jansson:  http://www.digip.org/jansson/ .  I suspect the python bindings would be Real Work, however.

**Comment by citrus-it** [comment]
> Probably not relevant here, but I was informed of a small C implementation called Jansson: http://www.digip.org/jansson/ . I suspect the python bindings would be Real Work, however.

tbh, the data need unserialising and that needs work and RAM. We tested in perl too (well, @hadfl did) and it that also uses approaching 200MiB of RAM to load and unserialise `catalog.summary.C`

`rapidjson` is an instant win (simplejson /stdlibjson compatible, backend in C++, nice clean implementation); there is just a lot of data in the OpenIndiana repository. Perhaps IPS could deal with it better, but that's not for this PR.

**Comment by hadfl** [comment]
Like @citrus-it  mentioned; I've tested pkg related JSON processing using perl with an established JSON module with a backend written in C. It only performed slightly better than rapidjson so I guess there is not too much there can be done regarding the JSON implementation.

And I am sure that pkg could be optimised here and there as well. But having said that (and the particular package was mentioned before), `consolidation/userland/userland-incorporation` ships in 300+ versions and each of it has 3k+ actions, so that will "blow up" to almost **1 million** strings in the data structure. I wonder if the best approach for the time being would be to check for packages with lots of actions and reduce the amount of different versions published in the repository.

**Comment by Toasterson** [comment]
We usually do a cleanup every other snapshot. But that  blocks older installations from installing packages. They need to upgrade before they can install packages again. @pyhalov has mentioned on IRC some times ago that maybe nuking `consolidation/userland/userland-incorporation` would be preferable.

File: src/modules/json.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/python3.7
```
Comment (pyhalov) [inline src/modules/json.py]: /usr/bin/python3.5

**Comment by pyhalov** [comment]
Test results of current IPS (without this patch):
```
======================================================================
BASELINE MISMATCH: The following results didn't match the baseline.
----------------------------------------------------------------------
cli.t_pkgrepo.py TestPkgRepo.test_10_list: pass
======================================================================
```

**Comment by pyhalov** [comment]
Test results of patched IPS (with this patch):
```
# Ran 1132 tests in 18539.442s - skipped 1 tests.
Skipped tests:
cli.t_pkg_help.py TestPkgHelp.test_help_character_encoding: Test Skipped: The test system must have the ja_JP.eucJP locale installed to run this test.


======================================================================
BASELINE MATCH
======================================================================
```

**Comment by pyhalov** [comment]
@citrus-it pkg should depend on rapidjson-35, fix src/pkg/manifests/package:pkg.p5m and src/pkg/manifests/developer:opensolaris:pkg5.p5m .

**Comment by citrus-it** [comment]
A couple of final benchmarks:

```
af@hipster:~$ time python3.5 /usr/bin/pkg list -g http://pkg.openindiana.org/hipster | wc -l
    4462

real    1m19.469s
user    1m7.112s
sys     0m4.449s

Peak RSS: 1107248K
```

and with rapidjson:

```
af@hipster:~/pkg5$ PYTHONPATH=`pwd`/proto/root_i386/usr/lib/python3.5/vendor-packages time python3.5 proto/root_i386/usr/bin/pkg list -g http://pkg.openindiana.org/hipster | wc -l

real       55.3
user       43.9
sys         3.7
    4462

Peak RSS:  929728K
```

File: src/pkg/manifests/package:pkg.p5m [hunk 2]
```diff
@@ -497,6 +498,6 @@ depend type=require fmri=library/python/prettytable-35
 depend type=require fmri=library/python/pybonjour-35
 depend type=require fmri=library/python/pycurl-35
 depend type=require fmri=library/python/pyopenssl-35
-depend type=require fmri=library/python/simplejson-35@3.6.5
+depend type=require fmri=library/python/rapidjson-35@3.6.5
```
Comment (pyhalov) [inline src/pkg/manifests/package:pkg.p5m]: rapidjson is at 0.9.1, please, remove version requirement

Comment (pyhalov) [reply src/pkg/manifests/package:pkg.p5m]: @citrus-it

Comment (citrus-it) [reply src/pkg/manifests/package:pkg.p5m]: Sorry, done!

**Comment by pyhalov** [comment]
Thanks

**Pull request merged**